languageServer/src/DvcYamlModel.ts

```diff
@@ -0,0 +1,177 @@
+  toString(): string
+}
+
+export class Cmd implements ICmd {
+  private cmd: Scalar | YAMLSeq
+
+  constructor(cmd: Scalar | YAMLSeq) {
+    this.cmd = cmd
+  }
+
+  static create(cmd: Scalar | YAMLSeq | undefined) {
+    return cmd ? new Cmd(cmd) : undefined
+  }
+
+  getReferencedFiles(): string[] {
+    const paths: string[] = []
+
+    if (isScalar(this.cmd)) {
+      const value = this.cmd.value as string
+      const matches = value.matchAll(/[\d/A-Za-z]+\.[A-Za-z]+/g)
+      for (const match of matches) {
+        paths.push(match[0])
+      }
+    }
+    return paths
+  }
+
+  toJSON(): JsonValue {
+    return this.cmd.toJSON()
+  }
+}
+
+export class Stage implements IDvcYamlStage {
+  name: string
+  cmd?: ICmd
+  deps?: Set<string>
+
+  constructor(private pairNode: Pair, public parentDvcYaml: IDvcYamlModel) {
+    this.name = pairNode.key as string
+    const value = pairNode.value
+
+    isMap(value) && this.buildModel(value)
+  }
+
+  buildModel(value: YAMLMap) {
+    const cmdNode = value.getIn(['cmd'], true) as Scalar | YAMLSeq | undefined
+    this.cmd = Cmd.create(cmdNode)
+
+    const depsNode = value.getIn(['deps'], true) as YAMLSeq | undefined
+    isSeq(depsNode) && this.buildDependencies(depsNode)
+  }
+
+  buildDependencies(depsNode: YAMLSeq) {
+    this.deps = new Set<string>()
+    for (const item of depsNode.items) {
+      if (isScalar(item)) {
+        this.deps.add(item.value as string)
+      }
+    }
+  }
+
+  toJSON(): JsonValue {
+    let json: Record<string, JsonValue> = {}
+    const nodeValue = this.pairNode.value
+
+    if (isNode(nodeValue)) {
+      json = {
+        ...nodeValue.toJSON()
+      }
+    }
+
+    if (this.cmd) {
+      json = {
+        ...json,
+        cmd: this.cmd.toJSON()
+      }
+    }
+
+    if (this.deps) {
+      json.deps = [...this.deps]
+    }
+
+    return json
+  }
+
+  addDependencies(items: string[]): void {
+    const current = this.deps || new Set<string>()
+
+    this.deps = new Set<string>([...current, ...items])
+  }
+
+  dependsOnAll(): boolean {
+    return true
+  }
+
+  contains(offset: number): boolean {
+    const valueNode = this.pairNode.value
+
+    return !!(
+      isNode(valueNode) &&
+      valueNode.range &&
+      inRange(offset, valueNode.range[0], valueNode.range[2])
+    )
+  }
+}
+
+export class DvcYaml implements IDvcYamlModel {
+  private stages: IDvcYamlStage[] = []
+
+  constructor(private document: IDvcTextDocument) {
+    const yamlDoc = this.document.getYamlDocument()
+    this.setStages(yamlDoc.get('stages', true))
+  }
+
+  setStages(node: unknown) {
+    if (isMap(node)) {
+      for (const pair of node.items) {
+        this.stages.push(new Stage(pair, this))
+      }
+    }
+  }
+
+  getStageAt(position: Position): IDvcYamlStage | undefined {
+    const offset = this.document.offsetAt(position)
+    return this.stages.find(stage => stage.contains(offset))
+  }
+
+  toString(): string {
```

Comment:
Can you please go through these classes and remove any code that is not currently being used to serve something to the client. We need to use YAGNI.